govcd/nsxv_distributed_firewall.go

```diff
@@ -0,0 +1,486 @@
+	if err != nil {
+		return nil, err
+	}
+
+	requestUrl, err := url.ParseRequestURI(initialUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	req := dfw.client.NewRequest(nil, http.MethodGet, *requestUrl, nil)
+
+	resp, err := checkResp(dfw.client.Http.Do(req))
+	if err != nil {
+		return nil, err
+	}
+	if resp != nil && resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("error fetching the service groups: %s", resp.Status)
+	}
+	var applicationGroupList types.ApplicationGroupList
+	err = decodeBody(types.BodyTypeXML, resp, &applicationGroupList)
+	if err != nil {
+		return nil, err
+	}
+	dfw.ServiceGroups = applicationGroupList.ApplicationGroup
+	return applicationGroupList.ApplicationGroup, nil
+}
+
+// Refresh retrieves fresh values for the distributed firewall rules, services, and service groups
+func (dfw *NsxvDistributedFirewall) Refresh() error {
+	if dfw.VdcId == "" {
+		return fmt.Errorf("no AdminVdc set for this NsxvDistributedFirewall")
+	}
+
+	_, err := dfw.GetServices(true)
+	if err != nil {
+		return err
+	}
+
+	_, err = dfw.GetServiceGroups(true)
+	if err != nil {
+		return err
+	}
+
+	_, err = dfw.GetConfiguration()
+	return err
+}
+
+// GetServiceById retrieves a single service, identified by its ID, for the current VDC
+// If the list of services was already retrieved, it uses it, otherwise fetches new ones.
+// Returns ErrorEntityNotFound when the requested services was not found
+func (dfw *NsxvDistributedFirewall) GetServiceById(serviceId string) (*types.Application, error) {
+	services, err := dfw.GetServices(false)
+	if err != nil {
+		return nil, err
+	}
+	for _, app := range services {
+		if app.ObjectID == serviceId {
+			return &app, nil
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetServiceByName retrieves a single service, identified by its name, for the current VDC
+// If the list of services was already retrieved, it uses it, otherwise fetches new ones.
+// Returns ErrorEntityNotFound when the requested service was not found
+func (dfw *NsxvDistributedFirewall) GetServiceByName(serviceName string) (*types.Application, error) {
+	services, err := dfw.GetServices(false)
+	if err != nil {
+		return nil, err
+	}
+	var foundService types.Application
+	for _, app := range services {
+		if app.Name == serviceName {
+			if foundService.ObjectID != "" {
+				return nil, fmt.Errorf("more than one service found with name '%s'", serviceName)
+			}
+			foundService = app
+		}
+	}
+	if foundService.ObjectID == "" {
+		return nil, ErrorEntityNotFound
+	}
+	return &foundService, nil
+}
+
+// GetServicesByRegex returns a list of services with their names matching the given regular expression
+// It may return an empty list (without error)
+func (dfw *NsxvDistributedFirewall) GetServicesByRegex(expression string) ([]types.Application, error) {
+	services, err := dfw.GetServices(false)
+	if err != nil {
+		return nil, err
+	}
+	searchRegex := regexp.MustCompile(expression)
```

Comment:
Similar to other place:
Given that this `expression` is supplied by user - IMO, we should use `regexp.Compile` and return the error to user if is invalid

Reply:
changed